libzmix/src/verifiable_encryption/cs_verifiable_encryption.rs

```diff
@@ -0,0 +1,849 @@
+        let u_sqr =
+            paillier_group.exponentiate(&ciphertext.u, &hs_x3_x2_times_2, Some(&mut ctx))?;
+        let v_sqr = paillier_group.sqr(&ciphertext.v, Some(&mut ctx))?;
+        if v_sqr != u_sqr {
+            return Err(UrsaCryptoError::from_msg(
+                UrsaCryptoErrorKind::InvalidStructure,
+                format!("u^2 != v^2, {:?} != {:?}", &u_sqr, &v_sqr),
+            ));
+        }
+        let mut messages = Vec::<BigNumber>::with_capacity(ciphertext.e.len());
+        for i in 0..ciphertext.e.len() {
+            let u_x1 =
+                paillier_group.exponentiate(&ciphertext.u, &pri_key.x1[i], Some(&mut ctx))?;
+            // 1/u^{x_1}
+            let u_x1_inv = u_x1.inverse(&paillier_group.modulus, Some(&mut ctx))?;
+            // (e/u^{x_1})
+            let e_u_x1_inv =
+                &ciphertext.e[i].mod_mul(&u_x1_inv, &paillier_group.modulus, Some(&mut ctx))?;
+
+            // m_hat = (e/u^{x_1})^2*t
+            let m_hat = paillier_group.exponentiate(
+                &e_u_x1_inv,
+                &pub_key.two_inv_times_2,
+                Some(&mut ctx),
+            )?;
+            if m_hat.modulus(&pub_key.n, Some(&mut ctx))? == *BIGNUMBER_1 {
+                let mut m = m_hat.modulus(&paillier_group.modulus, Some(&mut ctx))?;
+                m.sub_word(1)?;
+                m = m.div(&pub_key.n, Some(&mut ctx))?;
+                messages.push(m);
+            } else {
+                return Err(UrsaCryptoError::from_msg(
+                    UrsaCryptoErrorKind::InvalidStructure,
+                    format!("Decryption failed for message {}", i + 1),
+                ));
+            }
+        }
+
+        Ok(messages)
+    }
+
+    /// Encrypt multiple messages.
+    /// Encryption from section 3.2
+    pub fn encrypt(
+        messages: &[BigNumber],
+        label: &[u8],
+        pub_key: &CSEncPubkey,
+    ) -> UrsaCryptoResult<CSCiphertext> {
+        if messages.len() > pub_key.y1.len() {
+            return Err(UrsaCryptoError::from_msg(
+                UrsaCryptoErrorKind::InvalidStructure,
+                format!(
+                    "number of messages {} is more than supported by public key {}",
+                    messages.len(),
+                    pub_key.y1.len()
+                ),
+            ));
+        }
+
+        let paillier_group = &pub_key.paillier_group;
+        let r = paillier_group.rand_for_enc()?;
+
+        Self::encrypt_using_random_value(&r, messages, label, pub_key)
+    }
+
+    /// 1st phase of sigma protocol. Compute ciphertext and commitments (t values).
+    /// Return ciphertext, commitments and random values created during encryption and t value
+    /// "The protocol" from section 5.2. Not using t = g^m*h^s as the idemix protocol does not use it.
+    /// Guess is that since the knowledge of m is proved in the credential attribute proving protocol.
+    pub fn encrypt_and_prove_phase_1(
```

Comment:
Can this be wrapped in a transcript protocol where you just pass the transcript around and hide as much of the parameters as possible and it becomes a state machine? I have examples of this [here](https://github.com/mikelodder7/anoncreds-signatures/blob/master/src/bbs.rs#L207)

Reply:
The SignatureProtocol is composable as it assumes that it is the only protocol running. It computes challenge from the commitments it knows, but in practice, verifiable encryption is part of a bigger protocol so it needs to emit commitments (phase 1) and compute response (phase 3). The challenge comes from outside, maybe by hashing commitments from all sub protocols or maybe from verifier in case of an interactive ZKP. 
But we should probably define a SigmaProtocol interface that all other protocols could follow. Maybe borrow some stuff from [here](https://scapi.readthedocs.io/en/latest/interactive_layer/sigma_protocols.html). For managing the transcript, we can use merlin.

Reply:
Yes I was meaning the SignatureProtocol as an example. SigmaProtocol is fine.

Reply:
But can that SigmaProtocol protocol interface go in a separate PR? We will have to discuss about the interface anyway and that can take long.

Reply:
Okay as long as it doesn't get forgotten.